### PR TITLE
Remove OptionalCell's dependency on feature(const_fn) using where clauses

### PR DESCRIPTION
**Pull request by jrvanwhy**
### Pull Request Overview

This pull requests makes `OptionalCell` work with non-`Copy` types. This allows it to work without `#![feature(const_fn)]`. I rewrote a few methods that used `Cell::get` (which requires `T` to be `Copy`) that didn't need `T: Copy`.

`tock-cells` still needs `#![feature(const_fn)]` because `TakeCell` contains a mutable reference type, which cannot be initialized in `const` context without `const_fn`. Removing that dependency is possible but would require more `unsafe`, as we would need to replace `TockCell`'s reference with a raw pointer.

I have made the same semantic change using different syntax in #2010; I'd like input on which approach seems better.

This change helps migrate towards stable Rust (#1654).

### Testing Strategy

I commented out `#![feature(const_fn)]` and `pub mod take_cell;` in `libraries/tock-cells/src/lib.rs` then ran `cargo check` in `tock-cells` to verify I removed the dependency on `const_fn`. I reverted that change before committing.

I also ran `make allcheck` in the root of this repository.

### TODO or Help Wanted

I also sent #2010, which makes the same improvement in an equivalent but syntactically-different manner (using an `impl<T: Copy>` block rather than `where` clauses). I'd like feedback on which approach you find more maintainable.

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

File: libraries/tock-cells/src/optional_cell.rs [hunk 1]
```diff
@@ -55,38 +51,54 @@ impl<T: Copy> OptionalCell<T> {
...
+        self.value.set(value);
+        out
     }
 
     /// Returns the contained value or panics if contents is `None`.
-    pub fn expect(&self, msg: &str) -> T {
+    pub fn expect(&self, msg: &str) -> T
+    where
+        T: Copy,
+    {
```
Comment (hudson-ayers) [inline libraries/tock-cells/src/optional_cell.rs:83]: This still seems to require the contained type is `Copy`?

Comment (jrvanwhy) [reply libraries/tock-cells/src/optional_cell.rs:83]: Several methods of `OptionalCell` -- including `expect` -- require `T: Copy`.

Unless it panics, `expect` returns a copy of the value contained in the `OptionalCell` without modifying the `OptionalCell`, which definitely requires `T: Copy`.

Comment (hudson-ayers) [reply libraries/tock-cells/src/optional_cell.rs:83]: This makes sense, I was just confused by the first sentence of the PR "This pull requests makes OptionalCell work with non-Copy types"

**Comment by bradjc** [comment]
this: image

vs
#2010: image

I don't really have a preference.

**Comment by jrvanwhy** [comment]
I personally have a slight preference for #2010, as it more closely mirrors `core::cell::Cell`'s implementation.

**Pull request closed without merging**